tests/test_tpp_backend.py

```diff
@@ -1714,6 +1714,100 @@ def test_patients_address_as_of():
     )
 
 
+def test_date_window_behaviour_literals():
+    session = make_session()
+    session.add_all(
+        [
+            # This patient has event past midnight on end of date window
+            Patient(
+                CodedEvents=[
+                    CodedEvent(CTV3Code="foo1", ConsultationDate="2021-01-01"),
+                    CodedEvent(CTV3Code="foo2", ConsultationDate="2021-01-04"),
+                    CodedEvent(CTV3Code="foo3", ConsultationDate="2021-01-04T10:45:00"),
+                    CodedEvent(CTV3Code="foo2", ConsultationDate="2021-01-01T16:10:00"),
+                ]
+            ),
+            # This patient doesn't have any relevant events
+            Patient(
+                CodedEvents=[
+                    CodedEvent(CTV3Code="mto1", ConsultationDate="2010-01-01"),
+                    CodedEvent(CTV3Code="mto2", ConsultationDate="2010-01-14"),
+                    CodedEvent(CTV3Code="mto3", ConsultationDate="2010-01-20"),
+                    CodedEvent(CTV3Code="mto1", ConsultationDate="2010-02-04"),
+                    CodedEvent(CTV3Code="mto1", ConsultationDate="2012-01-01T10:45:00"),
+                    CodedEvent(CTV3Code="mtr2", ConsultationDate="2012-01-01T16:10:00"),
+                    CodedEvent(CTV3Code="mto1", ConsultationDate="2015-03-05"),
+                    CodedEvent(CTV3Code="mto1", ConsultationDate="2020-02-05"),
+                ]
+            ),
+        ]
+    )
+    session.commit()
+    foo_codes = codelist(["foo1", "foo2", "foo3"], "ctv3")
+    study = StudyDefinition(
+        population=patients.all(),
+        event_count=patients.with_these_clinical_events(
+            foo_codes,
+            between=["2021-01-01", "2021-01-04"],
+            returning="number_of_matches_in_period",
+        ),
+    )
+    results = study.to_dicts()
+    event_count = 4 if flags.WITH_END_DATE_FIX else 2
```

Comment:
I think that rather than changing the test assertion based on the flag, it would be better to keep the test the same but say that it's expected to fail if the fix isn't in place. You can do this with `@pytest.mark.xfail`:
https://docs.pytest.org/en/7.1.x/how-to/skipping.html#condition-parameter

The test will then still show as failing, but the failure won't cause the whole suite to fail.

Reply:
Just to explain a bit further: the thing which gave me pause here is that it just seems very odd to have tests which enforce known buggy behaviour. But it's quite common to have tests for desired behaviour that you known will currently fail, which is the purpose of the `xfail` marker.

It's not that I think anything bad will happen from the way things are currently done (especially given that the flag is presumably only temporary) it's just not very idomatic pytest so I thought it was worth mentioning.

Reply:
As discussed, you've explained that you want to assert that the tests give a very specific wrong result when the flag is disabled, rather than just failing in any old way — so it makes sense to have this slightly unusual style of test in this slightly unusual situation!